rfcs/20190610-standartizing-composite_ops.md

```diff
@@ -0,0 +1,349 @@
+tensorflow original implementation.
+
+Requires custom version of tensorflow to add new ops outside of tensorflow which
+makes it out of reach for most users and introduces incompatibilities within the
+ecosystem, effectively forcing large users to be stuck at old versions of tensorflow.
+
+#### Alternative 2: Use name scopes for delineation of the ops
+
+Pros:
+
+*   Simple and intuitive graph-def
+*   Backward compatible - no new operations are added.
+
+Cons:
+
+*   very brittle for the tools to optimize such ops. If they depend on scope
+    names it can easily cause conflicts with models that are doing something
+    else or accidental renames can cause scopes to become invisible for the
+    graph processing.
+*   If tools depend on graph pattern matching, this makes it hard to change
+    implementations later on.
+
+*   Tooling is not forward compatible.
+
+## Questions and Discussion Topics
+
+1.  We should provide namespacing for the operation names early on to avoid
+    future collisions with function names. One option is to adopt java-style
+    "org.opname", basically using the developing organization as disambiguating
+    factor. Another alternative is to use semantic namespaces e.g. `linalg`.  Yet
+    third, is to use org.semantic.opname.
+    
+    Note: since many of these functions will live in user codebase we obviously
+    can't and shouldn't police them, however having a recommended style guide
+    will simplify things later on.
+    
+    Should there be a `graduation` process? Where ops eventually move to a higher tier.
+    E.g. dedicated top-tier op like tensorflow.opname?
+    If so, we might also consider having another attribute 'aliases' to allow
+    history of names for backward compatibility.
+
+2.  If we go org/name, do we need centralized registry of accepted "org" names?
+
+3.  Do we need `reference` field? The idea is that points to the source of
+    authoritative ground truth implementation (likely not tensorflow
+    based), which this op is faithfully trying to reproduce? Should this
+    reference be structured? (For example it could point to existing
+    scipy/numpy/pillow library, or it could be pointing to a paper?). Should we
+    make this optional or get rid of it altogeher and delegate this to
+    documentation?
+    
+ 4. Name collisions - crash, no crash, etc. 
+ 
+ 5. tomhennigan: Aligning the definitions with MLIR 
+ 
+ I wonder if we should consider more metadata here. For example within a dialect MLIR op definitions [0] include a name [1], summary, description. Maybe we can align with them? Concretely I suggest considering something like:
+
+@tf.function(
+  op_def=tf.OpDef(
+      dialect="google",
+      name="mm_low_rank_matrix",
+      summary="..",
+      description="..",
+  ))
+def f(..):
+  pass
+[0] https://github.com/tensorflow/mlir/blob/master/g3doc/OpDefinitions.md#operation-definition
+[1] https://github.com/tensorflow/mlir/blob/master/g3doc/OpDefinitions.md#operation-name
```

Comment:
My understanding is that the "name" could match the MLIR operation name used by the downstream tool: so "google.mm_low_rank_matrix" should be enough. Since there should be a registration in MLIR for the operation itself, you shouldn't replicate here the summary and description though?